akka-http2-support/src/test/scala/akka/http/impl/engine/http2/PriorKnowledgeSwitchSpec.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.http.impl.engine.http2
+
+import scala.concurrent.Promise
+
+import akka.Done
+import akka.stream.ActorMaterializer
+import akka.stream.OverflowStrategy
+import akka.stream.QueueOfferResult.Enqueued
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.Keep
+import akka.stream.scaladsl.Source
+import akka.stream.scaladsl.Sink
+import akka.util.ByteString
+
+import akka.testkit.AkkaSpec
+import org.scalatest.exceptions.TestFailedException
+import org.scalatest.time.{ Span, Second, Milliseconds }
+
+class PriorKnowledgeSwitchSpec extends AkkaSpec {
```

Comment:
One good test to add here for regression testing purposes is an HTTP/1.0 request with no headers (HTTP/1.1 requires a Host header, so won't work). Eg:

```
GET / HTTP/1.0

```

The reason that test would be good is because that request is valid, and is shorter than the HTTP/2 preface. There's a possible optimisation that would break this request, that is to wait until you have enough bytes before comparing to the HTTP/2 preface, rather than doing the startsWith comparison on each set of bytes grabbed before the preface length is reached. With that optimisation in place, the above request will hang . A test case that tests this will ensure no one makes this optimisation in future.